### PR TITLE
Build FilteringPosterior on the fly

### PR DESCRIPTION
**Pull request by schmidtjonathan**
Builds the filtering posterior on the fly, i.e. directly builds up the `FilteringPosterior` object, which is returned in the end.
This happens by
- allowing for initializing an empty `TimeSeriesPosterior` object
- adding an `.append(loc, rv)` method to the `TimeSeriesPosterior` class

### CODE REVIEW

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (schmidtjonathan) [inline src/probnum/filtsmooth/timeseriesposterior.py:45]: @pnkraemer What do you think of this? This way, we save ourselves an additional `initialize` method that might add unnecessary complexity

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (schmidtjonathan) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:48]: @pnkraemer Do you approve of this change of the ordering of the arguments?
This is necessary in order to also create empty `FilteringPosterior` objects. Since non-optional arguments may not follow optional ones. And the transition **always** has to be provided. Even for empty posteriors.

**Comment by schmidtjonathan** [comment]
Also, do we even need another method? As is now, we provide
- `.filter()` for when you want just the end result of the Kalman filter on the entire time period, which now builds on-the-fly, but gains no novel functionalities from this PR
- `filtered_states_generator()` which I did not touch in this PR, that can be wrapped from outside, in case somebody wants to investigate their posterior in every step.

So, somebody who wants to create e.g. an animation, could do sth like this:

```python
posterior = FilteringPosterior(transition=self.dynamics_model)

for idx, (rv, _) in enumerate(
  self.filtered_states_generator(regression_problem, _previous_posterior)
):
  posterior.append(location=regression_problem.locations[idx], state=rv)
  plot_posterior(posterior)
```
, right?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:48]: Yes, I could live with that ;) However, do you think it would improve readability even more if at initialisation one would (need to) pass the empty lists? This might make it even more obvious to a code reader what happens, but that is just a guess. What do you think?

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/timeseriesposterior.py:45]: yes, I am fine with that. When I mentioned the `initialise` methods, I meant one in the filter (not in the posterior). See also my other comment :)

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (mmahsereci) [inline src/probnum/filtsmooth/timeseriesposterior.py:62]: this accesses a private class.

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (mmahsereci) [inline src/probnum/filtsmooth/timeseriesposterior.py:52]: Is this TODO still open?

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/timeseriesposterior.py:62]: see the "hence the underscore" bit in #388

Comment (mmahsereci) [reply src/probnum/filtsmooth/timeseriesposterior.py:62]: thanks

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:48]: I personally don't think that it would improve readability, or ease-of-use. But that's personal taste :)

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/timeseriesposterior.py:52]: Yep.
Technically, we should check the types here, because it wouldn't throw an error if somebody appends all kidns of stuff to those lists.
But this method is probably invoked very often during filtering. So, I wanted to raise the question whether or not we want type checks here?
Safety vs. speed, yet again.

Comment (schmidtjonathan) [reply src/probnum/filtsmooth/timeseriesposterior.py:52]: What do you think, @mmahsereci , @pnkraemer ?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:200]: do you think the generator should yield t as well? This enumerate seems a bit awkward here, doesnt it?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:48]: alright, then not :)

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/timeseriesposterior.py:52]: I think its fine like this. An extra type check seems a bit unnecessary to me...

Comment (schmidtjonathan) [reply src/probnum/filtsmooth/timeseriesposterior.py:52]: Ok thanks to both of you :)

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:200]: Done

**Comment by schmidtjonathan** [comment]
@pnkraemer I resolved what we discussed. But we should discuss whether we should merge this PR **after** merging #396 . I would say _yes_, because this here has way less changes and might be less confusing to integrate.

**Comment by pnkraemer** [comment]
That would be wonderful. Thank you for suggesting that.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalman.py:262]: sorry if this has been explained before.
Can you please remind me what the reason for returning the time value here was? It seems like this value would be accessible outside of the generator, but I might be missing something.

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (pnkraemer) [inline src/probnum/filtsmooth/timeseriesposterior.py:59]: do you think it makes sense to cache those? And does it also make sense to use (i) check in the append() whether the location is right of the last location (ii) add a finalise() method that blocks the append() for the future? I am not sure whether these things are good or bad ideas, but would like to hear your opinion :)

**Review by pnkraemer — COMMENTED** [review]
I have two minor questions reg. the design of the functions. Mostly a brainstorm. When these  things are resolved (and the branch is up to date again), we can merge :)

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py]: It is returned such that when building the `FilteringPosterior` on the fly, we can call `posterior.append(location=t, state=rv)`.
Do you think there's a more elegant solution?

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (schmidtjonathan) [reply src/probnum/filtsmooth/timeseriesposterior.py]: > do you think it makes sense to cache those? 

I don't know what happens then when the object is altered (say, by appending another location+state)

(i): 
Mh technically, I would agree. But isn't that checked elsewhere? Resp., does the algorithm, which calls this method, even allow this?

(ii)
Absolutely, good idea. I'll add that. Do we want another class (e.g. `FrozenFilteringPosterior` ) for that? Or just a flag?

File: src/probnum/filtsmooth/gaussfiltsmooth/kalman.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalman.py]: not sure. I guess the alternative would be to include zip over `regression_problem.locations, filter_generator(regression_problem)` but (a) this looks awkward, and (b) the current implementation *might* be more robust w.r.t. multiple measurements at a single locations.

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/timeseriesposterior.py]: > I don't know what happens then when the object is altered (say, by appending another location+state)
I guess it remains cached. But I am not 100% sure (though, isn't the idea of caching to not check these things anymore?)

(I) the method does not do this, but what happens if a user randomly decides to append a new measurement. This would be possible (though stupid? I don't know). I don't think it is really checked, but perhaps there are some weird errors. I don't know. I think it would be reasonable to check.

(ii) I think a flag would work, unless the FrozenFilteringPosterior makes the code more elegant. Do what you consider most simple :)

Comment (schmidtjonathan) [reply src/probnum/filtsmooth/timeseriesposterior.py]: I used a flag now, since I didn't find an elegant way to transfer the changes of a `FrozenTimeSeriesPosterior` to subclasses of `TimeSeriesPosterior` , like `FilteringPosterior` for instance. Before creating a hundred subclasses, I chose the easier way. Although it would have been nice to convert the property into a cached_property after freezing.

**Comment by schmidtjonathan** [comment]
@pnkraemer Ready for review.

File: src/probnum/filtsmooth/timeseriesposterior.py
Comment (pnkraemer) [reply src/probnum/filtsmooth/timeseriesposterior.py]: Alright! 

> Although it would have been nice to convert the property into a cached_property after freezing

Technically, one can always manually cache it :) But I think it is okay without

Comment (schmidtjonathan) [reply src/probnum/filtsmooth/timeseriesposterior.py]: Only if you know beforehand how many states/locations you will add. But that's a good point, we could make that optionally possible. Shall we?

Sth like
```python
def __init__(..., total=None):
  if total is not None:
    self.locations = np.empty((total, ))
  ...
  # etc.
```

Comment (schmidtjonathan) [reply src/probnum/filtsmooth/timeseriesposterior.py]: Don't know how much we gain by this, though. I suspect not much, tbh

**Review by pnkraemer — APPROVED** [review]

**Pull request merged**
